pkg/controller/tf/controller.go

```diff
@@ -164,6 +164,28 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (res
 		r.logger.Info("Skipping reconcile as nothing has changed and 0 reconcile period is set", "resource", req.NamespacedName)
 		return reconcile.Result{}, nil
 	}
+
+	cc, ccc, err := resourceactuation.FetchLiveKCCState(ctx, r.mgr.GetClient(), req.NamespacedName)
```

Comment:
We are doing some generic processing of the KRM object through out the reconciliation. Such processing includes applying the default values for KCC owned annotation fields, ensuring the finalizers are properly set, etc. I'm not concerned about the defaulting logic, but I think we'll probably want to ensure the finalizers are set properly: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/tf/controller.go#L311. Finalizers we have are used to help manage the deletion: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/deletiondefender/controller.go#L99

This reminds me of another tricky edge case(s) we need to deal with: What if the defaulting logic changes over versions? This can be between KCC versions (should be rare because it is backwards incompatible) or CRD versions (safer, but still will cause confusions).

E.g. in CC/CCC v1beta1, the default value for `state-into-spec` is `merge`, and in CC/CCC v1beta2, the default value for `state-into-spec` will be `absent`. If the users installs KCC via CC/CCC v1beta1 in two clusters, one reconciling and the other paused, upgrades CC/CCC to v1beta2 at some points, and then needs to unpause the paused cluster, they will suddenly see some behavioral changes.

This use case is something I'm concerned about in the unpausing scenario that we will probably need to think carefully about later.

Reply:
Thanks for highlighting this! I think the concerns are well founded.

I moved this code block around and now call `EnsureFinalizers` as part of handling the Paused scenario. 

I think the general idea is that we want to keep reconciling the KRM resource but not actually actuate it onto the cloud provider. So by that logic, all other KRM specific that would happen as part of `sync()`  should probably be factored out so we can apply them in the Paused case too. At first glance, EnsureFinalizers is the only thing that we should definitely include. I'm open to revisit that and keep factoring things out as necessary.

Reply:
No problem! Most of them can be addressed/improved in the future.

Have you verified that EnsureFinalizers runs properly? I had some problems testing in #1179 so just want to confirm it is indeed working.